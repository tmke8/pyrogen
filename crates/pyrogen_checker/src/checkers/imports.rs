use std::path::Path;

use pyrogen_python_ast::{imports::ImportMap, PySourceType};
use pyrogen_python_index::Indexer;
use pyrogen_source_file::Locator;
use rustpython_ast::Suite;

use crate::{registry::Diagnostic, settings::CheckerSettings, source_kind::SourceKind};

#[allow(clippy::too_many_arguments)]
pub(crate) fn check_imports(
    python_ast: &Suite,
    locator: &Locator,
    indexer: &Indexer,
    settings: &CheckerSettings,
    path: &Path,
    package: Option<&Path>,
    source_kind: &SourceKind,
    source_type: PySourceType,
) -> (Vec<Diagnostic>, Option<ImportMap>) {
    todo!("check_imports")
}
