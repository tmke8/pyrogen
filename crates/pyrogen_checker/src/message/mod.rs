use std::cmp::Ordering;
use std::collections::BTreeMap;
use std::io::Write;
use std::ops::Deref;

use rustc_hash::FxHashMap;
use rustpython_parser::ast::Ranged;
use rustpython_parser::text_size::{TextRange, TextSize};

use pyrogen_source_file::{SourceFile, SourceLocation};

pub use github::GithubEmitter;
pub use json::JsonEmitter;
pub use text::TextEmitter;

use crate::registry::{Diagnostic, DiagnosticKind};
use crate::settings::code_table::MessageKind;

// mod diff;
mod github;
mod json;
mod text;

#[derive(Debug, PartialEq, Eq)]
pub struct Message {
    pub diagnostic: DiagnosticKind,
    pub range: TextRange,
    pub file: SourceFile,
    pub ignore_offset: TextSize,
    pub kind: MessageKind,
}

impl Message {
    pub fn from_diagnostic(
        diagnostic: Diagnostic,
        file: SourceFile,
        ignore_offset: TextSize,
        kind: MessageKind,
    ) -> Self {
        Self {
            range: diagnostic.range(),
            diagnostic: diagnostic.kind,
            file,
            ignore_offset,
            kind,
        }
    }

    pub fn filename(&self) -> &str {
        self.file.name()
    }

    pub fn compute_start_location(&self) -> SourceLocation {
        self.file.to_source_code().source_location(self.start())
    }

    pub fn compute_end_location(&self) -> SourceLocation {
        self.file.to_source_code().source_location(self.end())
    }
}

impl Ord for Message {
    fn cmp(&self, other: &Self) -> Ordering {
        (&self.file, self.start()).cmp(&(&other.file, other.start()))
    }
}

impl PartialOrd for Message {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ranged for Message {
    fn range(&self) -> TextRange {
        self.range
    }
}

struct MessageWithLocation<'a> {
    message: &'a Message,
    start_location: SourceLocation,
}

impl Deref for MessageWithLocation<'_> {
    type Target = Message;

    fn deref(&self) -> &Self::Target {
        self.message
    }
}

fn group_messages_by_filename(messages: &[Message]) -> BTreeMap<&str, Vec<MessageWithLocation>> {
    let mut grouped_messages = BTreeMap::default();
    for message in messages {
        grouped_messages
            .entry(message.filename())
            .or_insert_with(Vec::new)
            .push(MessageWithLocation {
                message,
                start_location: message.compute_start_location(),
            });
    }
    grouped_messages
}

/// Display format for a [`Message`]s.
///
/// The emitter serializes a slice of [`Message`]'s and writes them to a [`Write`].
pub trait Emitter {
    /// Serializes the `messages` and writes the output to `writer`.
    fn emit(&mut self, writer: &mut dyn Write, messages: &[Message]) -> anyhow::Result<()>;
}

#[cfg(test)]
mod tests {
    use rustc_hash::FxHashMap;
    use rustpython_parser::ast::Ranged;
    use rustpython_parser::text_size::{TextRange, TextSize};

    use pyrogen_source_file::SourceFileBuilder;

    use crate::message::{Emitter, Message};
    use crate::registry::{Diagnostic, DiagnosticKind, ErrorCode};
    use crate::settings::code_table::MessageKind;

    pub(super) fn create_messages() -> Vec<Message> {
        let fib = r#"import os


def fibonacci(n):
    """Compute the nth number in the Fibonacci sequence."""
    x = 1
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)
"#;

        let unused_import = Diagnostic::new(
            DiagnosticKind {
                error_code: ErrorCode::UnusedImport,
                body: "`os` imported but unused".to_string(),
            },
            TextRange::new(TextSize::from(7), TextSize::from(9)),
        );

        let fib_source = SourceFileBuilder::new("fib.py", fib).finish();

        let unused_variable = Diagnostic::new(
            DiagnosticKind {
                error_code: ErrorCode::UnusedVariable,
                body: "Local variable `x` is assigned to but never used".to_string(),
            },
            TextRange::new(TextSize::from(94), TextSize::from(95)),
        );

        let file_2 = r#"if a == 1: pass"#;

        let undefined_name = Diagnostic::new(
            DiagnosticKind {
                error_code: ErrorCode::UndefinedName,
                body: "Undefined name `a`".to_string(),
            },
            TextRange::new(TextSize::from(3), TextSize::from(4)),
        );

        let file_2_source = SourceFileBuilder::new("undef.py", file_2).finish();

        let unused_import_start = unused_import.start();
        let unused_variable_start = unused_variable.start();
        let undefined_name_start = undefined_name.start();
        vec![
            Message::from_diagnostic(
                unused_import,
                fib_source.clone(),
                unused_import_start,
                MessageKind::Warning,
            ),
            Message::from_diagnostic(
                unused_variable,
                fib_source,
                unused_variable_start,
                MessageKind::Warning,
            ),
            Message::from_diagnostic(
                undefined_name,
                file_2_source,
                undefined_name_start,
                MessageKind::Error,
            ),
        ]
    }

    pub(super) fn capture_emitter_output(
        emitter: &mut dyn Emitter,
        messages: &[Message],
    ) -> String {
        let mut output: Vec<u8> = Vec::new();
        emitter.emit(&mut output, messages).unwrap();

        String::from_utf8(output).expect("Output to be valid UTF-8")
    }
}
