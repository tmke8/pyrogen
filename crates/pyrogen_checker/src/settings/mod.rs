use std::path::{Path, PathBuf};

use crate::{registry::Rule, settings::types::PythonVersion};
use path_absolutize::path_dedot;
use pyrogen_macros::CacheKey;

use self::rule_table::RuleTable;

pub mod flags;
pub mod rule_table;
pub mod types;

#[derive(Debug, CacheKey)]
pub struct CheckerSettings {
    pub project_root: PathBuf,
    pub rules: RuleTable,

    pub target_version: PythonVersion,
    pub namespace_packages: Vec<PathBuf>,
    pub src: Vec<PathBuf>,
}

impl CheckerSettings {
    pub fn new(project_root: &Path) -> Self {
        Self {
            target_version: PythonVersion::default(),
            project_root: project_root.to_path_buf(),
            rules: RuleTable::from_iter(vec![Rule::SyntaxError].into_iter()),
            namespace_packages: vec![],

            src: vec![path_dedot::CWD.clone()],
        }
    }

    #[must_use]
    pub fn with_target_version(mut self, target_version: PythonVersion) -> Self {
        self.target_version = target_version;
        self
    }
}
