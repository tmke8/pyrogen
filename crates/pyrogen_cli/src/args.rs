use std::path::PathBuf;

use clap::{command, Parser};

use pyrogen_checker::logging::LogLevel;
use pyrogen_checker::settings::types::{FilePattern, PythonVersion};
use pyrogen_workspace::configuration::Configuration;
use pyrogen_workspace::resolver::ConfigurationTransformer;

#[derive(Debug, Parser)]
#[command(
    author,
    name = "pyrogen",
    about = "Pyrogen, an extremely strict Python type checker."
)]
#[command(version)]
pub struct Args {
    #[clap(flatten)]
    pub checker_args: CheckCommand,
    #[clap(flatten)]
    pub log_level_args: LogLevelArgs,
}

// The `Parser` derive is for pyrogen_dev, for pyrogen_cli `Args` would be sufficient
#[derive(Clone, Debug, clap::Parser)]
#[allow(clippy::struct_excessive_bools)]
pub struct CheckCommand {
    /// List of files or directories to check.
    pub files: Vec<PathBuf>,
    /// The minimum Python version that should be supported.
    #[arg(long, value_enum)]
    pub target_version: Option<PythonVersion>,
    /// Path to the `pyproject.toml` or `pyrogen.toml` file to use for
    /// configuration.
    #[arg(long, conflicts_with = "isolated")]
    pub config: Option<PathBuf>,
    /// List of paths, used to omit files and/or directories from analysis.
    #[arg(
        long,
        value_delimiter = ',',
        value_name = "FILE_PATTERN",
        help_heading = "File selection"
    )]
    pub exclude: Option<Vec<FilePattern>>,
    /// Respect file exclusions via `.gitignore` and other standard ignore files.
    /// Use `--no-respect-gitignore` to disable.
    #[arg(
        long,
        overrides_with("no_respect_gitignore"),
        help_heading = "File selection"
    )]
    respect_gitignore: bool,
    #[clap(long, overrides_with("respect_gitignore"), hide = true)]
    no_respect_gitignore: bool,
    /// Enforce exclusions, even for paths passed to Ruff directly on the command-line.
    /// Use `--no-force-exclude` to disable.
    #[arg(
        long,
        overrides_with("no_force_exclude"),
        help_heading = "File selection"
    )]
    force_exclude: bool,
    #[clap(long, overrides_with("force_exclude"), hide = true)]
    no_force_exclude: bool,
    /// Disable cache reads.
    #[arg(short, long, help_heading = "Miscellaneous")]
    pub no_cache: bool,
    /// Ignore all configuration files.
    #[arg(long, conflicts_with = "config", help_heading = "Miscellaneous")]
    pub isolated: bool,
    /// Path to the cache directory.
    #[arg(long, env = "PYROGEN_CACHE_DIR", help_heading = "Miscellaneous")]
    pub cache_dir: Option<PathBuf>,
    /// The name of the file when passing it through stdin.
    #[arg(long, help_heading = "Miscellaneous")]
    pub stdin_filename: Option<PathBuf>,
    /// Exit with status code "0", even upon detecting lint violations.
    #[arg(short, long, help_heading = "Miscellaneous")]
    pub exit_zero: bool,
}

#[allow(clippy::module_name_repetitions)]
#[derive(Debug, clap::Args)]
pub struct LogLevelArgs {
    /// Enable verbose logging.
    #[arg(
        short,
        long,
        global = true,
        group = "verbosity",
        help_heading = "Log levels"
    )]
    pub verbose: bool,
    /// Print diagnostics, but nothing else.
    #[arg(
        short,
        long,
        global = true,
        group = "verbosity",
        help_heading = "Log levels"
    )]
    pub quiet: bool,
    /// Disable all logging (but still exit with status code "1" upon detecting diagnostics).
    #[arg(
        short,
        long,
        global = true,
        group = "verbosity",
        help_heading = "Log levels"
    )]
    pub silent: bool,
}

impl From<&LogLevelArgs> for LogLevel {
    fn from(args: &LogLevelArgs) -> Self {
        if args.silent {
            Self::Silent
        } else if args.quiet {
            Self::Quiet
        } else if args.verbose {
            Self::Verbose
        } else {
            Self::Default
        }
    }
}

impl CheckCommand {
    /// Partition the CLI into command-line arguments and configuration
    /// overrides.
    pub fn partition(self) -> (CheckArguments, CliOverrides) {
        (
            CheckArguments {
                config: self.config,
                exit_zero: self.exit_zero,
                files: self.files,
                no_cache: self.no_cache,
                isolated: self.isolated,
                stdin_filename: self.stdin_filename,
            },
            CliOverrides {
                exclude: self.exclude,
                respect_gitignore: resolve_bool_arg(
                    self.respect_gitignore,
                    self.no_respect_gitignore,
                ),
                target_version: self.target_version,
                // TODO(charlie): Included in `pyproject.toml`, but not inherited.
                cache_dir: self.cache_dir,
                force_exclude: resolve_bool_arg(self.force_exclude, self.no_force_exclude),
            },
        )
    }
}

fn resolve_bool_arg(yes: bool, no: bool) -> Option<bool> {
    match (yes, no) {
        (true, false) => Some(true),
        (false, true) => Some(false),
        (false, false) => None,
        (..) => unreachable!("Clap should make this impossible"),
    }
}

/// CLI settings that are distinct from configuration (commands, lists of files,
/// etc.).
#[allow(clippy::struct_excessive_bools)]
pub struct CheckArguments {
    pub config: Option<PathBuf>,
    pub exit_zero: bool,
    pub files: Vec<PathBuf>,
    pub isolated: bool,
    pub no_cache: bool,
    pub stdin_filename: Option<PathBuf>,
}

/// CLI settings that function as configuration overrides.
#[derive(Clone, Default)]
#[allow(clippy::struct_excessive_bools)]
pub struct CliOverrides {
    pub exclude: Option<Vec<FilePattern>>,
    pub respect_gitignore: Option<bool>,
    pub target_version: Option<PythonVersion>,
    // TODO(charlie): Captured in pyproject.toml as a default, but not part of `Settings`.
    pub cache_dir: Option<PathBuf>,
    pub force_exclude: Option<bool>,
}

impl ConfigurationTransformer for CliOverrides {
    fn transform(&self, mut config: Configuration) -> Configuration {
        if let Some(cache_dir) = &self.cache_dir {
            config.cache_dir = Some(cache_dir.clone());
        }
        if let Some(exclude) = &self.exclude {
            config.exclude = Some(exclude.clone());
        }
        if let Some(force_exclude) = &self.force_exclude {
            config.force_exclude = Some(*force_exclude);
        }
        if let Some(respect_gitignore) = &self.respect_gitignore {
            config.respect_gitignore = Some(*respect_gitignore);
        }
        if let Some(target_version) = &self.target_version {
            config.target_version = Some(*target_version);
        }

        config
    }
}
