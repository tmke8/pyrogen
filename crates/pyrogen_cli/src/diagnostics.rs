#![cfg_attr(target_family = "wasm", allow(dead_code))]

use std::io;
use std::ops::AddAssign;
#[cfg(unix)]
use std::os::unix::fs::PermissionsExt;
use std::path::Path;

use anyhow::{Context, Result};
use colored::Colorize;
use filetime::FileTime;
use log::{debug, error, warn};
use rustpython_parser::text_size::{TextRange, TextSize};
use thiserror::Error;

use pyrogen_checker::checker::{lint_only, CheckerResult};
use pyrogen_checker::fs;
use pyrogen_checker::logging::DisplayParseError;
use pyrogen_checker::message::Message;
use pyrogen_checker::pyproject_toml::lint_pyproject_toml;
use pyrogen_checker::registry::{AsErrorCode, Diagnostic, DiagnosticKind, ErrorCode};
use pyrogen_checker::settings::{flags, CheckerSettings};
use pyrogen_checker::source_kind::SourceKind;
use pyrogen_macros::CacheKey;
use pyrogen_python_ast::imports::ImportMap;
use pyrogen_python_ast::{SourceType, TomlSourceType};
use pyrogen_source_file::{LineIndex, SourceCode, SourceFileBuilder};
use pyrogen_workspace::Settings;

use crate::cache::Cache;

#[derive(CacheKey)]
pub(crate) struct FileCacheKey {
    /// Timestamp when the file was last modified before the (cached) check.
    file_last_modified: FileTime,
    /// Permissions of the file before the (cached) check.
    file_permissions_mode: u32,
}

impl FileCacheKey {
    fn from_path(path: &Path) -> io::Result<FileCacheKey> {
        // Construct a cache key for the file
        let metadata = path.metadata()?;

        #[cfg(unix)]
        let permissions = metadata.permissions().mode();
        #[cfg(windows)]
        let permissions: u32 = metadata.permissions().readonly().into();

        Ok(FileCacheKey {
            file_last_modified: FileTime::from_last_modification_time(&metadata),
            file_permissions_mode: permissions,
        })
    }
}

#[derive(Debug, Default, PartialEq)]
pub(crate) struct Messages {
    pub(crate) messages: Vec<Message>,
    pub(crate) imports: ImportMap,
}

impl Messages {
    pub(crate) fn new(messages: Vec<Message>, imports: ImportMap) -> Self {
        Self { messages, imports }
    }

    /// Generate [`Messages`] based on a [`SourceExtractionError`].
    pub(crate) fn from_source_error(
        err: &SourceExtractionError,
        path: Option<&Path>,
        settings: &CheckerSettings,
    ) -> Self {
        let diagnostic = Diagnostic::from(err);
        if let Some(kind) = settings.table.entry(diagnostic.kind.error_code()) {
            let name = path.map_or_else(|| "-".into(), std::path::Path::to_string_lossy);
            let dummy = SourceFileBuilder::new(name, "").finish();
            Self::new(
                vec![Message::from_diagnostic(
                    diagnostic,
                    dummy,
                    TextSize::default(),
                    kind,
                )],
                ImportMap::default(),
            )
        } else {
            match path {
                Some(path) => {
                    warn!(
                        "{}{}{} {err}",
                        "Failed to check ".bold(),
                        fs::relativize_path(path).bold(),
                        ":".bold()
                    );
                }
                None => {
                    warn!("{}{} {err}", "Failed to check".bold(), ":".bold());
                }
            }

            Self::default()
        }
    }
}

impl AddAssign for Messages {
    fn add_assign(&mut self, other: Self) {
        self.messages.extend(other.messages);
        self.imports.extend(other.imports);
    }
}

/// Type-check the source code at the given `Path`.
pub(crate) fn type_check_path(
    path: &Path,
    package: Option<&Path>,
    settings: &CheckerSettings,
    cache: Option<&Cache>,
    respect_type_ignore: flags::TypeIgnore,
) -> Result<Messages> {
    // Check the cache.
    let caching = match cache {
        Some(cache) if respect_type_ignore.into() => {
            let relative_path = cache
                .relative_path(path)
                .expect("wrong package cache for file");

            let cache_key = FileCacheKey::from_path(path).context("Failed to create cache key")?;

            if let Some(cache) = cache.get(relative_path, &cache_key) {
                return Ok(cache.as_diagnostics(path));
            }

            // Stash the file metadata for later so when we update the cache it reflects the prerun
            // information
            Some((cache, relative_path, cache_key))
        }
        _ => None,
    };

    debug!("Checking: {}", path.display());

    let source_type = match SourceType::from(path) {
        SourceType::Toml(TomlSourceType::Pyproject) => {
            let messages = if settings
                .table
                .iter_enabled()
                .any(|rule_code| rule_code.lint_source().is_pyproject_toml())
            {
                let contents =
                    match std::fs::read_to_string(path).map_err(SourceExtractionError::Io) {
                        Ok(contents) => contents,
                        Err(err) => {
                            return Ok(Messages::from_source_error(&err, Some(path), settings));
                        }
                    };
                let source_file = SourceFileBuilder::new(path.to_string_lossy(), contents).finish();
                lint_pyproject_toml(source_file, settings)
            } else {
                vec![]
            };
            return Ok(Messages {
                messages,
                ..Messages::default()
            });
        }
        SourceType::Toml(_) => return Ok(Messages::default()),
        SourceType::Python(source_type) => source_type,
    };

    // Extract the sources from the file.
    let LintSource(source_kind) = match LintSource::try_from_path(path) {
        Ok(Some(sources)) => sources,
        Ok(None) => return Ok(Messages::default()),
        Err(err) => {
            return Ok(Messages::from_source_error(&err, Some(path), settings));
        }
    };
    let source_kind = SourceKind::new(source_kind);

    // Lint the file.
    let CheckerResult {
        data: (messages, imports),
        error: parse_error,
    } = lint_only(
        path,
        package,
        settings,
        respect_type_ignore,
        &source_kind,
        source_type,
    );

    let imports = imports.unwrap_or_default();

    if let Some((cache, relative_path, key)) = caching {
        // We don't cache parsing errors.
        if parse_error.is_none() {
            cache.update(relative_path.to_owned(), key, &messages, &imports);
        }
    }

    if let Some(err) = parse_error {
        error!(
            "{}",
            DisplayParseError::new(
                err,
                SourceCode::new(
                    source_kind.source_code(),
                    &LineIndex::from_source_text(source_kind.source_code())
                ),
                &source_kind,
            )
        );
    }

    Ok(Messages { messages, imports })
}

/// Generate `Diagnostic`s from source code content derived from
/// stdin.
pub(crate) fn type_check_stdin(
    path: Option<&Path>,
    package: Option<&Path>,
    contents: String,
    settings: &Settings,
    noqa: flags::TypeIgnore,
) -> Result<Messages> {
    let SourceType::Python(source_type) = path.map(SourceType::from).unwrap_or_default() else {
        return Ok(Messages::default());
    };

    let source_kind = SourceKind::new(contents);

    // Lint the inputs.
    let CheckerResult {
        data: (messages, imports),
        error: parse_error,
    } = lint_only(
        path.unwrap_or_else(|| Path::new("-")),
        package,
        &settings.checker,
        noqa,
        &source_kind,
        source_type,
    );

    let imports = imports.unwrap_or_default();

    if let Some(err) = parse_error {
        error!(
            "Failed to parse {}: {err}",
            path.map_or_else(|| "-".into(), fs::relativize_path).bold()
        );
    }

    Ok(Messages { messages, imports })
}

#[derive(Debug)]
pub(crate) struct LintSource(String);

impl LintSource {
    /// Extract the lint [`LintSource`] from the given file path.
    pub(crate) fn try_from_path(path: &Path) -> Result<Option<LintSource>, SourceExtractionError> {
        // This is tested by ruff_cli integration test `unreadable_file`
        let contents = std::fs::read_to_string(path)?;
        Ok(Some(LintSource(contents)))
    }
}

#[derive(Error, Debug)]
pub(crate) enum SourceExtractionError {
    /// The extraction failed due to an [`io::Error`].
    #[error(transparent)]
    Io(#[from] io::Error),
}

impl From<&SourceExtractionError> for Diagnostic {
    fn from(err: &SourceExtractionError) -> Self {
        match err {
            // IO errors.
            SourceExtractionError::Io(_) => Diagnostic::new(
                DiagnosticKind {
                    error_code: ErrorCode::IOError,
                    body: err.to_string(),
                },
                TextRange::default(),
            ),
        }
    }
}
