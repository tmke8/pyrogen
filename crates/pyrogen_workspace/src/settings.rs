use std::path::{Path, PathBuf};

use path_absolutize::path_dedot;
use pyrogen_cache::cache_dir;
use pyrogen_checker::settings::{
    types::{FilePattern, FilePatternSet},
    CheckerSettings,
};
use pyrogen_macros::CacheKey;

#[derive(Debug, CacheKey)]
#[allow(clippy::struct_excessive_bools)]
pub struct Settings {
    #[cache_key(ignore)]
    pub cache_dir: PathBuf,
    pub file_resolver: FileResolverSettings,

    pub checker: CheckerSettings,
}

impl Default for Settings {
    fn default() -> Self {
        let project_root = path_dedot::CWD.as_path();
        Self {
            cache_dir: cache_dir(project_root),
            checker: CheckerSettings::new(project_root),
            file_resolver: FileResolverSettings::new(project_root),
        }
    }
}

pub(crate) static EXCLUDE: &[FilePattern] = &[
    FilePattern::Builtin(".bzr"),
    FilePattern::Builtin(".direnv"),
    FilePattern::Builtin(".eggs"),
    FilePattern::Builtin(".git"),
    FilePattern::Builtin(".git-rewrite"),
    FilePattern::Builtin(".hg"),
    FilePattern::Builtin(".ipynb_checkpoints"),
    FilePattern::Builtin(".mypy_cache"),
    FilePattern::Builtin(".nox"),
    FilePattern::Builtin(".pants.d"),
    FilePattern::Builtin(".pyenv"),
    FilePattern::Builtin(".pytest_cache"),
    FilePattern::Builtin(".pytype"),
    FilePattern::Builtin(".pyrogen_cache"),
    FilePattern::Builtin(".svn"),
    FilePattern::Builtin(".tox"),
    FilePattern::Builtin(".venv"),
    FilePattern::Builtin(".vscode"),
    FilePattern::Builtin("__pypackages__"),
    FilePattern::Builtin("_build"),
    FilePattern::Builtin("buck-out"),
    FilePattern::Builtin("build"),
    FilePattern::Builtin("dist"),
    FilePattern::Builtin("node_modules"),
    FilePattern::Builtin("venv"),
];

pub(crate) static INCLUDE: &[FilePattern] = &[
    FilePattern::Builtin("*.py"),
    FilePattern::Builtin("*.pyi"),
    FilePattern::Builtin("**/pyproject.toml"),
];

#[derive(Debug, CacheKey)]
pub struct FileResolverSettings {
    pub exclude: FilePatternSet,
    pub force_exclude: bool,
    pub include: FilePatternSet,
    pub respect_gitignore: bool,
    pub project_root: PathBuf,
}

impl FileResolverSettings {
    fn new(project_root: &Path) -> Self {
        Self {
            project_root: project_root.to_path_buf(),
            exclude: FilePatternSet::try_from_iter(EXCLUDE.iter().cloned()).unwrap(),
            force_exclude: false,
            respect_gitignore: true,
            include: FilePatternSet::try_from_iter(INCLUDE.iter().cloned()).unwrap(),
        }
    }
}
