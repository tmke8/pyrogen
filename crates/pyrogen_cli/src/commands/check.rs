use std::collections::HashMap;
use std::fmt::Write;
use std::io;
use std::path::{Path, PathBuf};
use std::time::Instant;

use anyhow::Result;
use colored::Colorize;
use ignore::Error;
use itertools::Itertools;
use log::{debug, error, warn};
#[cfg(not(target_family = "wasm"))]
use rayon::prelude::*;
use rustc_hash::FxHashMap;
use rustpython_parser::text_size::{TextRange, TextSize};

use pyrogen_checker::message::Message;
use pyrogen_checker::registry::ErrorCode;
use pyrogen_checker::settings::{flags, CheckerSettings};
use pyrogen_checker::{fs, warn_user_once};
use pyrogen_python_ast::imports::ImportMap;
use pyrogen_source_file::SourceFileBuilder;
use pyrogen_workspace::resolver::{
    python_files_in_path, PyprojectConfig, PyprojectDiscoveryStrategy,
};

use crate::args::CliOverrides;
use crate::cache::{self, Cache};
use crate::diagnostics::Diagnostics;
use crate::panic::catch_unwind;

/// Run the checker over a collection of files.
pub(crate) fn check(
    files: &[PathBuf],
    pyproject_config: &PyprojectConfig,
    overrides: &CliOverrides,
    cache: flags::Cache,
    noqa: flags::Noqa,
) -> Result<Diagnostics> {
    // Collect all the Python files to check.
    let start = Instant::now();
    let (paths, resolver) = python_files_in_path(files, pyproject_config, overrides)?;
    let duration = start.elapsed();
    debug!("Identified files to lint in: {:?}", duration);

    if paths.is_empty() {
        warn_user_once!("No Python files found under the given path(s)");
        return Ok(Diagnostics::default());
    }

    // Initialize the cache.
    if cache.into() {
        fn init_cache(path: &Path) {
            if let Err(e) = cache::init(path) {
                error!("Failed to initialize cache at {}: {e:?}", path.display());
            }
        }

        match pyproject_config.strategy {
            PyprojectDiscoveryStrategy::Fixed => {
                init_cache(&pyproject_config.settings.cache_dir);
            }
            PyprojectDiscoveryStrategy::Hierarchical => {
                for settings in
                    std::iter::once(&pyproject_config.settings).chain(resolver.settings())
                {
                    init_cache(&settings.cache_dir);
                }
            }
        }
    };

    // Discover the package root for each Python file.
    let package_roots = resolver.package_roots(
        &paths
            .iter()
            .flatten()
            .map(ignore::DirEntry::path)
            .collect::<Vec<_>>(),
        pyproject_config,
    );

    // Load the caches.
    let caches = bool::from(cache).then(|| {
        package_roots
            .iter()
            .map(|(package, package_root)| package_root.unwrap_or(package))
            .unique()
            .par_bridge()
            .map(|cache_root| {
                let settings = resolver.resolve(cache_root, pyproject_config);
                let cache = Cache::open(cache_root.to_path_buf(), settings);
                (cache_root, cache)
            })
            .collect::<HashMap<&Path, Cache>>()
    });

    let start = Instant::now();
    let mut diagnostics: Diagnostics = paths
        .par_iter()
        .map(|entry| {
            match entry {
                Ok(entry) => {
                    let path = entry.path();
                    let package = path
                        .parent()
                        .and_then(|parent| package_roots.get(parent))
                        .and_then(|package| *package);

                    let settings = resolver.resolve(path, pyproject_config);

                    let cache_root = package.unwrap_or_else(|| path.parent().unwrap_or(path));
                    let cache = caches.as_ref().and_then(|caches| {
                        if let Some(cache) = caches.get(&cache_root) {
                            Some(cache)
                        } else {
                            debug!("No cache found for {}", cache_root.display());
                            None
                        }
                    });

                    lint_path(path, package, &settings.checker, cache, noqa).map_err(|e| {
                        (Some(path.to_owned()), {
                            let mut error = e.to_string();
                            for cause in e.chain() {
                                write!(&mut error, "\n  Cause: {cause}").unwrap();
                            }
                            error
                        })
                    })
                }
                Err(e) => Err((
                    if let Error::WithPath { path, .. } = e {
                        Some(path.clone())
                    } else {
                        None
                    },
                    e.io_error()
                        .map_or_else(|| e.to_string(), io::Error::to_string),
                )),
            }
            .unwrap_or_else(|(path, message)| {
                if let Some(path) = &path {
                    let settings = resolver.resolve(path, pyproject_config);
                    warn!(
                        "{}{}{} {message}",
                        "Failed to lint ".bold(),
                        fs::relativize_path(path).bold(),
                        ":".bold()
                    );
                    Diagnostics::default()
                } else {
                    warn!("{} {message}", "Encountered error:".bold());
                    Diagnostics::default()
                }
            })
        })
        .reduce(Diagnostics::default, |mut acc, item| {
            acc += item;
            acc
        });

    diagnostics.messages.sort();

    // Store the caches.
    if let Some(caches) = caches {
        caches
            .into_par_iter()
            .try_for_each(|(_, cache)| cache.store())?;
    }

    let duration = start.elapsed();
    debug!("Checked {:?} files in: {:?}", paths.len(), duration);

    Ok(diagnostics)
}

/// Wraps [`lint_path`](crate::diagnostics::lint_path) in a [`catch_unwind`](std::panic::catch_unwind) and emits
/// a diagnostic if the linting the file panics.
fn lint_path(
    path: &Path,
    package: Option<&Path>,
    settings: &CheckerSettings,
    cache: Option<&Cache>,
    noqa: flags::Noqa,
) -> Result<Diagnostics> {
    let result =
        catch_unwind(|| crate::diagnostics::lint_path(path, package, settings, cache, noqa));

    match result {
        Ok(inner) => inner,
        Err(error) => {
            let message = r#"This indicates a bug in Pyrogen. If you could open an issue at:

    https://github.com/tmke8/pyrogen/issues/new?title=%5BChecker%20panic%5D

...with the relevant file contents, the `pyproject.toml` settings, and the following stack trace, we'd be very appreciative!
"#;

            error!(
                "{}{}{} {message}\n{error}",
                "Panicked while linting ".bold(),
                fs::relativize_path(path).bold(),
                ":".bold()
            );

            Ok(Diagnostics::default())
        }
    }
}

#[cfg(test)]
#[cfg(unix)]
mod test {
    use std::fs;
    use std::os::unix::fs::OpenOptionsExt;

    use anyhow::Result;
    use rustc_hash::FxHashMap;
    use tempfile::TempDir;

    use pyrogen_checker::message::{Emitter, TextEmitter};
    use pyrogen_checker::registry::ErrorCode;
    use pyrogen_checker::settings::{flags, CheckerSettings};
    use pyrogen_workspace::resolver::{PyprojectConfig, PyprojectDiscoveryStrategy};
    use pyrogen_workspace::Settings;

    use crate::args::CliOverrides;

    use super::check;

    /// We check that regular python files, pyproject.toml and jupyter notebooks all handle io
    /// errors gracefully
    #[test]
    fn unreadable_files() -> Result<()> {
        let path = "E902.py";
        let rule_code = ErrorCode::IOError;

        // Create inaccessible files
        let tempdir = TempDir::new()?;
        let pyproject_toml = tempdir.path().join("pyproject.toml");
        let python_file = tempdir.path().join("code.py");
        for file in [&pyproject_toml, &python_file] {
            fs::OpenOptions::new()
                .create(true)
                .write(true)
                .mode(0o000)
                .open(file)?;
        }

        // Configure
        let snapshot = format!("{}_{}", rule_code.to_str(), path);
        // invalid pyproject.toml is not active by default
        let settings = Settings {
            checker: CheckerSettings::for_rules(vec![rule_code, ErrorCode::InvalidPyprojectToml]),
            ..Settings::default()
        };
        let pyproject_config =
            PyprojectConfig::new(PyprojectDiscoveryStrategy::Fixed, settings, None);

        // Run
        let diagnostics = check(
            // Notebooks are not included by default
            &[tempdir.path().to_path_buf()],
            &pyproject_config,
            &CliOverrides::default(),
            flags::Cache::Disabled,
            flags::Noqa::Disabled,
        )
        .unwrap();
        let mut output = Vec::new();

        TextEmitter::default()
            .with_show_fix_status(true)
            .emit(&mut output, &diagnostics.messages)
            .unwrap();

        let messages = String::from_utf8(output).unwrap();

        insta::with_settings!({
            omit_expression => true,
            filters => vec![
                // The tempdir is always different (and platform dependent)
                (tempdir.path().to_str().unwrap(), "/home/ferris/project"),
            ]
        }, {
            insta::assert_snapshot!(snapshot, messages);
        });
        Ok(())
    }
}
