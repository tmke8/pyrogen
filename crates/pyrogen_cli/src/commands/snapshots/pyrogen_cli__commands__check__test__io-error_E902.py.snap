---
source: crates/pyrogen_cli/src/commands/check.rs
---
/home/ferris/project/code.py:1:1: error: Permission denied (os error 13) [io-error]
/home/ferris/project/pyproject.toml:1:1: error: Permission denied (os error 13) [io-error]

